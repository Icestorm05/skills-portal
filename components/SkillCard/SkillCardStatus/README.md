# Skill Card Status

## Functionality

The skill card status displays a readonly list of the level and currency passed in.

## Usage

### Props

`<skill-card-status>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `currency` | Number OR Null *optional* | The skill currency ID. Defaults to `null`.
| `level` | Number OR Null *optional* | The skill level ID. Defaults to `null`.

### Example

```html
<skill-card-status :currency="4"
                   :level="2"/>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import SkillCardStatus from '~/components/SkillCard/SkillCardStatus/SkillCardStatus.vue';

@Component({
    components: {
        SkillCardStatus,
    },
})
export default class GenericComponent extends Vue {}
```