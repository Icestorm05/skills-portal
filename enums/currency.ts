export enum ECurrency {
  Historic = 4,
  Fading = 3,
  Current = 2
}
