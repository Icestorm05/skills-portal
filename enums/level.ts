import theme from "~/assets/json/theme.json";

export enum ELevel {
  Awareness = 2,
  Practitioner = 3,
  Expert = 4
}

export const LevelColors = {
  4: theme.success,
  3: theme.warning,
  2: theme.error
};
