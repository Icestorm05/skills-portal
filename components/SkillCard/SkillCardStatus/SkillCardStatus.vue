<template lang="pug">
    v-layout(full-width).SkillCardStatus.text-xs-left
        v-flex
            p(v-if="level")
                b Level: 
                v-chip(:color="levelColor"
                       :text-color="levelTextColor"
                       small).SkillCardLevelStatus {{ levels[level] }}
            p(v-if="currency")
                b Currency: 
                v-chip(:color="currencyColor"
                       :text-color="currencyTextColor"
                       small).SkillCardCurrencyStatus {{ currencies[currency] }}
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import theme from '~/assets/json/theme.json';

import { ECurrency } from "~/enums/currency";
import { ELevel } from "~/enums/level";

import CalcTextColor from "~/mixins/CalcTextColor";

import VChip from "vuetify/es5/components/VChip";
import { VFlex, VLayout } from "vuetify/es5/components/VGrid";

@Component({
  components: {
    VChip,
    VFlex,
    VLayout
  },
  mixins: [CalcTextColor]
})
export default class SkillCardStatus extends CalcTextColor {
  @Prop({
    default: null,
    validator: val => val === null || typeof val === "number"
  })
  private currency: number | null;
  @Prop({
    default: null,
    validator: val => val === null || typeof val === "number"
  })
  private level: number | null;

  private currencies = ECurrency;
  private levels = ELevel;

  private get currencyColor() {
    if (this.currency === 2) {
      return "primary";
    } else if (this.currency === 3) {
      return "warning";
    } else if (this.currency === 4) {
      return "error";
    }
  }

  private get currencyTextColor() {
    return this.currencyColor && this.textColor(theme[this.currencyColor]) === "#ffffff"
      ? "white"
      : "black";
  }

  private get levelColor() {
    if (this.level === 2) {
      return "error";
    } else if (this.level === 3) {
      return "warning";
    } else if (this.level === 4) {
      return "success";
    }
  }

  private get levelTextColor() {
    return this.levelColor && this.textColor(theme[this.levelColor]) === "#ffffff"
      ? "white"
      : "black";
  }
}
</script>
