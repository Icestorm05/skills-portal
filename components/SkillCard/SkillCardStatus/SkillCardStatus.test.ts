import SkillCardStatus from "./SkillCardStatus.vue";
import { mount } from "@vue/test-utils";

import { ECurrency } from "~/enums/currency";
import { ELevel } from "~/enums/level";

declare const expect: jest.Expect;

describe("Skill Card Status", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(SkillCardStatus);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", async () => {
    const wrapper = mount(SkillCardStatus, {
      propsData: { currency: 2, level: 4 }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("currency and level prop are null by default", () => {
    const wrapper = mount(SkillCardStatus);
    expect(wrapper.vm.$props.currency).toEqual(null);
    expect(wrapper.vm.$props.level).toEqual(null);
  });

  test("currency and level prop can be set to null", () => {
    const wrapper = mount(SkillCardStatus, {
      propsData: { currency: null, level: null }
    });
    expect(wrapper.vm.$props.currency).toEqual(null);
    expect(wrapper.vm.$props.level).toEqual(null);
  });

  test("setting currency and level prop overrides default", () => {
    const wrapper = mount(SkillCardStatus, {
      propsData: { currency: 2, level: 4 }
    });
    expect(wrapper.vm.$props.currency).toEqual(2);
    expect(wrapper.vm.$props.level).toEqual(4);
  });

  test("the component by default is blank", () => {
    const wrapper = mount(SkillCardStatus);
    expect(wrapper.find(".SkillCardLevelStatus").exists()).toBeFalsy();
    expect(wrapper.find(".SkillCardCurrencyStatus").exists()).toBeFalsy();
  });

  test("setting level prop displays the level paragraph", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 2 } });
    const level = wrapper.find(".SkillCardLevelStatus");
    expect(level.exists()).toBeTruthy();
    expect(level.text()).toEqual(ELevel[2]);
  });

  test("level paragraph updates with prop change", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 2 } });
    const level = wrapper.find(".SkillCardLevelStatus");
    wrapper.setProps({ level: 3 });
    expect(level.text()).toEqual(ELevel[3]);
  });

  test("setting currency prop displays currency paragraph", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 4 } });
    const currency = wrapper.find(".SkillCardCurrencyStatus");
    expect(currency.exists()).toBeTruthy();
    expect(currency.text()).toEqual(ECurrency[4]);
  });

  test("currency paragraph updates with prop change", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 4 } });
    const currency = wrapper.find(".SkillCardCurrencyStatus");
    wrapper.setProps({ currency: 3 });
    expect(currency.text()).toEqual(ECurrency[3]);
  });

  test("currency color getter returns primary with currency set to 2", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 2 } });
    expect((wrapper.vm as any).currencyColor).toEqual("primary");
  });

  test("currency color getter returns warning with currency set to 3", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 3 } });
    expect((wrapper.vm as any).currencyColor).toEqual("warning");
  });

  test("currency color getter returns error with currency set to 4", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 4 } });
    expect((wrapper.vm as any).currencyColor).toEqual("error");
  });

  test("currency color getter updates with prop change", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { currency: 4 } });
    wrapper.setProps({ currency: 2 });
    expect((wrapper.vm as any).currencyColor).toEqual("primary");
  });

  test("currency color getter returns undefined with null currency", () => {
    const wrapper = mount(SkillCardStatus);
    expect((wrapper.vm as any).currencyColor).toBeUndefined();
  });

  test("level color getter returns error with level set to 2", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 2 } });
    expect((wrapper.vm as any).levelColor).toEqual("error");
  });

  test("level color getter returns warning with level set to 3", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 3 } });
    expect((wrapper.vm as any).levelColor).toEqual("warning");
  });

  test("level color getter returns success with level set to 4", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 4 } });
    expect((wrapper.vm as any).levelColor).toEqual("success");
  });

  test("level color getter updates with prop change", () => {
    const wrapper = mount(SkillCardStatus, { propsData: { level: 4 } });
    wrapper.setProps({ level: 2 });
    expect((wrapper.vm as any).levelColor).toEqual("error");
  });

  test("level color getter returns undefined with null level", () => {
    const wrapper = mount(SkillCardStatus);
    expect((wrapper.vm as any).levelColor).toBeUndefined();
  });
});
