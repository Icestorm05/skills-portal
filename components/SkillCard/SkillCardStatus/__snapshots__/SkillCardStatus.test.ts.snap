// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill Card Status snapshot has not changed 1`] = `
<div class="layout SkillCardStatus text-xs-left full-width">
    <div class="flex">
        <p>
            <b>Level: </b><span tabindex="0" class="chip SkillCardLevelStatus chip--small success white--text"><span class="chip__content">Expert</span></span>
        </p>
        <p>
            <b>Currency: </b><span tabindex="0" class="chip SkillCardCurrencyStatus chip--small primary white--text"><span class="chip__content">Current</span></span>
        </p>
    </div>
</div>
`;
